backend/lucene/src/main/java/org/hibernate/search/backend/lucene/search/predicate/impl/AbstractLuceneSearchNestedPredicateBuilder.java

```diff
@@ -6,23 +6,59 @@
  */
 package org.hibernate.search.backend.lucene.search.predicate.impl;
 
+import java.util.LinkedList;
+import java.util.List;
+
+import org.hibernate.search.util.common.AssertionFailure;
+
 import org.apache.lucene.search.Query;
 
 public abstract class AbstractLuceneSearchNestedPredicateBuilder extends AbstractLuceneSearchPredicateBuilder {
 
-	private final String nestedDocumentPath;
+	private final List<String> nestedPathHierarchy;
 
-	public AbstractLuceneSearchNestedPredicateBuilder(String nestedDocumentPath) {
-		this.nestedDocumentPath = nestedDocumentPath;
+	public AbstractLuceneSearchNestedPredicateBuilder(List<String> nestedPathHierarchy) {
+		this.nestedPathHierarchy = nestedPathHierarchy;
 	}
 
 	@Override
 	public final Query build(LuceneSearchPredicateContext context) {
-		if ( nestedDocumentPath == null || nestedDocumentPath.equals( context.getNestedPath() ) ) {
+		List<String> nestedSteps = implicitNestedSteps( context );
+		return applyImplicitNestedSteps( nestedSteps, context );
+	}
+
+	private List<String> implicitNestedSteps(LuceneSearchPredicateContext context) {
+		String contextNestedPath = context.getNestedPath();
+
+		if ( contextNestedPath == null ) {
+			// we need to handle all the nestedPathHierarchy belong to the target
+			return new LinkedList<>( nestedPathHierarchy );
+		}
+
+		int contextNestedPathIndex = nestedPathHierarchy.indexOf( contextNestedPath );
+		if ( contextNestedPathIndex < 0 ) {
+			throw new AssertionFailure( "The nested path must belong to the nested path hierarchy: there's an Hibernate Search bug." );
+		}
+
+		// we need to handle just the last part of the nestedPathHierarchy belong to the target,
+		// the one that hasn't been handled by the context
+		return new LinkedList<>( nestedPathHierarchy.subList( contextNestedPathIndex + 1, nestedPathHierarchy.size() ) );
+	}
+
+	private Query applyImplicitNestedSteps(List<String> furtherImplicitNestedSteps, LuceneSearchPredicateContext context) {
+		if ( furtherImplicitNestedSteps.isEmpty() ) {
 			return super.build( context );
 		}
 
-		LuceneSearchPredicateContext childContext = new LuceneSearchPredicateContext( nestedDocumentPath );
-		return LuceneNestedPredicateBuilder.doBuild( context, nestedDocumentPath, super.build( childContext ) );
+		if ( furtherImplicitNestedSteps.size() == 1 ) {
+			String lastStep = furtherImplicitNestedSteps.get( 0 );
+			LuceneSearchPredicateContext childContext = new LuceneSearchPredicateContext( lastStep );
+			// super.build( childContext ) must be called only from the very last nest step
+			return LuceneNestedPredicateBuilder.doBuild( context, lastStep, super.build( childContext ) );
+		}
+
+		String step = furtherImplicitNestedSteps.remove( 0 );
+		LuceneSearchPredicateContext childContext = new LuceneSearchPredicateContext( step );
+		return LuceneNestedPredicateBuilder.doBuild( context, step, applyImplicitNestedSteps( furtherImplicitNestedSteps, childContext ) );
```

Comment:
Same comment as for the Elasticsearch backend: a loop would be much clearer than recursion here.